### ISSUE TITLE
JacksonModule not ThreadSafe

### ISSUE THREAD

**Issue by kkondratov** [victools/jsonschema-generator#249]
Apologies for some vagueness, this is due to corporate code, i will try to elaborate the situation as much as possible.

Jackson module uses a HashMap for beanDescriptions when resolving beanDescriptionsForClass.

The use of computeIfAbsent with a normal hashmap causes ConcurrentModificationException when adding standardDefinition to the context.

In our implementation we provide a CustomSchemaDefinitionProviderV2 to the SchemaGeneratorConfigBuilder, which in provideCustomDefinition method adds a standardDefinition by calling `context.createStandardDefinition` in certain cases.
I think this causes the internal maps to be updated causing the exception

I added the stacktrace stripped from the corporate code, the comment line in there is the call to the `context.createStandardDefinition`

```
at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1221)
at com.github.victools.jsonschema.module.jackson.JacksonModule.getBeanDescriptionForClass(JacksonModule.java:222)
at com.github.victools.jsonschema.module.jackson.JacksonModule.shouldIgnoreField(JacksonModule.java:239)
at com.github.victools.jsonschema.generator.SchemaGeneratorConfigPart.lambda$shouldIgnore$0(SchemaGeneratorConfigPart.java:129)
at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
at java.base/java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:632)
at com.github.victools.jsonschema.generator.SchemaGeneratorConfigPart.shouldIgnore(SchemaGeneratorConfigPart.java:129)
at com.github.victools.jsonschema.generator.impl.SchemaGeneratorConfigImpl.shouldIgnore(SchemaGeneratorConfigImpl.java:287)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.lambda$collectFields$4(SchemaGenerationContextImpl.java:504)
at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178)
at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.collectFields(SchemaGenerationContextImpl.java:505)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.collectObjectProperties(SchemaGenerationContextImpl.java:460)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.generateObjectDefinition(SchemaGenerationContextImpl.java:417)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.addSubtypeReferencesInDefinition(SchemaGenerationContextImpl.java:332)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.traverseGenericType(SchemaGenerationContextImpl.java:305)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.createStandardDefinition(SchemaGenerationContextImpl.java:200)
///// REMOVED LINE Due to corporate -> previously explained call done here
at com.github.victools.jsonschema.generator.impl.SchemaGeneratorConfigImpl.lambda$getCustomDefinition$1(SchemaGeneratorConfigImpl.java:243)
at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
at java.base/java.util.ArrayList$SubList$2.tryAdvance(ArrayList.java:1458)
at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
at com.github.victools.jsonschema.generator.impl.SchemaGeneratorConfigImpl.getCustomDefinition(SchemaGeneratorConfigImpl.java:245)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.traverseGenericType(SchemaGenerationContextImpl.java:266)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.traverseGenericType(SchemaGenerationContextImpl.java:243)
at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.parseType(SchemaGenerationContextImpl.java:96)
at com.github.victools.jsonschema.generator.SchemaBuilder.createSchemaForSingleType(SchemaBuilder.java:111)
at com.github.victools.jsonschema.generator.SchemaBuilder.createSingleTypeSchema(SchemaBuilder.java:56)
at com.github.victools.jsonschema.generator.SchemaGenerator.generateSchema(SchemaGenerator.java:60)
```

**Comment by CarstenWickner** [comment]
Hi @kkondratov,

Three questions:
1. Are you adding anything to the standard definition after creating it?
    - If not, perhaps `context.createStandardDefinitionReference()` is the more appropriate call for you.
    - If yes, then please make sure you're providing the `CustomSchemaDefinitionProviderV2` instance as parameter to the `context.createStandardDefinition()` call.
2. What version of the `jsonschema-generator` dependencies are you using? The `JacksonModule` line `222` indicates you're not on the latest release, as that no longer corresponds to the `getBeanDescriptionForClass()` method.
3. Are you creating multiple schemas in parallel? The `JacksonModule` is not threadsafe (as of today), but for single-threaded usage, there should be no `ConcurrentModificationException`. I suspect you're responsible for the concurrency yourself.

The task for me is straightforward: making it all thread-safe. But I cannot promise when I'll get to that.
However, the workaround should also be simple enough: accept that you cannot concurrently generate multiple schemas with the same `JacksonModule` instance – at least not with the current release.

**Comment by kkondratov** [comment]
Hi @CarstenWickner,

Thank you for a quick reply and a thorough analysis.

To answer your questions:

1. No we are not adding anything to the standard definition, we are returning a custom definition like this:
`new CustomDefinition(context.createStandardDefinition(javaType, this), true)` 
2. We're using version 4.17.0 of `jsonschema-generator`
3. That might be the case, I am not the original author of this piece of code thus i think the assumption was made here that `JacksonModule` was threadsafe (seeing that Jackson ObjectMapper is). We will definitely address this on our side as well.

My reasoning for submitting the ticket is to let point out that the JacksonModule is not threadsafe. I was not expecting a quick fix or had any reason to apply pressure for this to be fixed in a quick manner. 

We will address it on our end as well to prevent further errors.

Once again thank you for a clear and thorough reply/analysis and your work regarding `jsonschema-generator`.

**Comment by CarstenWickner** [comment]
Fixed release as part of [v4.24.3](https://github.com/victools/jsonschema-generator/releases/tag/v4.24.3).
